Present the Fallthrough typeclass as an alternative to PartialService

kryptt opened the pull request:
A work in progress; servers need to be aware of the fallthrough just like bryce's take.

bryce-anderson commented on server/src/main/scala/org/http4s/server/package.scala:
```diff
@@ -30,43 +59,19 @@ package object server {
+      * Allows Service chainig through an implicit [[Fallthrough]] instance.
+      *
+      */
+    def orElse[A, B : Fallthrough](a: Service[A, B], b: Service[A, B]): Service[A, B] =
```
'The people' will want this as an infix operator, and I think that makes sense. Obviously an explicit function is good too.

rossabaker replied:
I had imagined it taking a middleware-like syntax where the last param list was the Service to modify:

```scala
def withFallback[A, B: Fallthrough](fallback: Service[A, B])(service: Service[A, B]): Service[A, B]
```

For the infix syntax we all know and love:

```scala
implicit class ServiceSyntax[A, B](service: Service[A, B]) {
  def orElse(fallback: Service[A, B]) = withFallback(fallback)(service)
}
```

bryce-anderson commented on server/src/main/scala/org/http4s/server/package.scala:
```diff
@@ -82,8 +87,7 @@ package object server {
      * with status [[Status.NotFound]] for any requests where the function
      * is undefined.
      */
-    def apply(pf: PartialFunction[Request, Task[Response]]): HttpService =
```
This constructor is there largely because of the core dsl's propensity to generate `PartialFunction`s under common (and I think reasonable) usage. Is there a good reason to remove it?

ShengC replied:
+1 I would keep this method. A partial function is essentially the most convenient way to build up routes using `dsl`.
I would also add an arg `default` to this method to give user a chance to override the default behavior (which most likely will be a `404`) if the partial function does not match a `Request`

rossabaker replied:
:+1:, and the default param is also a nice idea.

bryce-anderson commented:
I know this hasn't been addressed in this PR yet, but with `Fallthrough` being an opt in mechanism, does it make sense to mandate it for the servers, eg, the `Router`? I'm willing to bet it could be made opt-in as well but the form seems somewhat strange in my imagination.

rossabaker commented on server/src/main/scala/org/http4s/server/package.scala:
```diff
@@ -1,13 +1,44 @@
+    * For any given B, if a Fallthrough[B] exists within implicit context
+    * then [[Service#orElse]] can be used.
+    */
+  trait Fallthrough[B] {
```
This may be the old Java programmer in me talking, and totally doesn't matter, but I see Fallthrough and companion in its own file.

rossabaker commented on server/src/main/scala/org/http4s/server/package.scala:
```diff
@@ -1,13 +1,44 @@
+    */
+  trait Fallthrough[B] {
+    def isFallthrough(a: B): Boolean
+    def fallthrough[A](fst: B, snd: Service[A, B]): Service[A, B] =
```
A `Service` needs to know about `Fallthrough`, but `Fallthrough` doesn't need to know anything about `Service`.  Same package, so not a real problematic cyclic dependency... I just expected this to live elsewhere.

rossabaker commented:
I made some bikesheddy comments, but I like the direction here.

kryptt commented:
It's been a hectic morning, I'll get back on this later today

ShengC commented on server/src/main/scala/org/http4s/server/package.scala:
```diff
@@ -1,13 +1,44 @@
+    implicit def forResponse: Fallthrough[Response] = new Fallthrough[Response] {
+      def isFallthrough(r: Response): Boolean = r.status.code == 404
+    }
+
```
should we put something like this in the companion as well ? I think it'd be useful for debug.

    def stay: Fallthrough[Response] = new Fallthrough[Response] { def isFallthrough(r: Response): Boolean = false }

rossabaker replied:
That seems fine.

Are we now agreeing that 404, vs. something with an attribute, is the default implicit fallthrough Response?

kryptt commented:
Ok, so I implemented all of your suggestions and rebased the commits.

Router has been re-written, the .sort is still needed since we are running routes by prefix length instead of definition order (in any case the sort is run just once on service definition).

Now, all tests pass including the ones added by @bryce-anderson in #389 and RouterSpec#match longer prefixes first.

kryptt commented on server/src/test/scala/org/http4s/server/HttpServiceSpec.scala:
```diff
@@ -0,0 +1,40 @@
+      aggregate1.apply(Request(uri = uri("/wontMatch"))).run.as[String].run must equal ("srvc2")
+    }
+
+    "Safely skip undefined behavior" in {
```
This is different from #389 in that any custom 404 is treated as fall through. 

Personally, I believe honoring custom 404 over non-custom ones under specific situations could lead to unexpected results on how fall through services would be evaluated...

rossabaker replied:
I think this debate is a big motivating factor to having `Fallthrough` be an implicit which can be made explicit under a different definition.  I do agree that this choice is the least surprising default definition.

kryptt replied:
:+1: With 8dea190 we now also exposes the Fallthrough implicit when defining routes.

rossabaker commented:
:+1:

ShengC commented:
One more thing - do we really a `Monoid` constraint on `FallThrough` ? From what I can read, `fallThrough` operation only needs to know if `B` as in `Service[A, B]` can be `Zero` or Empty, and does not require `B` should be composable at all.

kryptt commented:
@ShengC : Fallthough is lawless and you need to implement is: `def isFallthrough(a: B): Boolean`

What you see there are two default implementations. One for `B : Monoid : Equal` and another for `HttpService`, together they should allow for 95% of the usecases to *just work* while allowing anyone to provide their own implementation (for example one where custom 404 responses have a higher priority than `HttpService.notFound`.

ShengC commented:
@kryptt 

    def fallthrough[A](fst: B, snd: Service[A, B]): Service[A, B] =
      if (isFallthrough(fst)) snd else Service.constVal(fst)

in order to let `fallThrough` work, I suppose one only needs to be specific about `isFallthrough`. In case of `HttpService`, that is defined as `_.status == 404` . In case of `Monoid` based one, that is defined as `_ == Zero`. My point is you don't need a `Monoid` constraint to make up `isFallthrough`, instead you only need something that _can be_ zero or empty.

rossabaker commented:
`Monoid` overconstrains things, but it's the simplest thing in the standard libraries that introduces a zero.  @kryptt has worked on a [Unital typeclass](https://github.com/kryptt/scalaz-outlaws/commit/d5182b7a9df9a968c614fab12456e321f5e7d397) for scalaz-outlaws.  In Cats, it's `alleycats.Empty`.  Each of these is a dependency that I'd rather not take on, especially when the `Fallthrough[Response]` will by far be the most commonly used parameterization.

kryptt commented:
@rossabaker : yes.

Guys, I think this is pretty much done, unless someone finds anything I glaringly missed. If this is the direction we want to go,  by all means merge it in.

(Is the test failure a travis hickup? all tests run successfully here)

rossabaker commented:
Yes, that's a spurious Travis failure.

:+1: to this as is.

bryce-anderson commented:
I think this PR is strictly better than what we have in master and am cool with merging it. :+1: I like the typeclass solution for allowing arbitrary `orElse` behavior while still providing the obvious default (fallthrough on 404).

This may be somewhat off topic for this PR, and I'm sure everyone knows my stance already, I think the quest to recycle similar FP types is leading to unnecessary obfuscation of what it is required in order to start a http4s server. The cabal of constructs which includes a type alias to a refined `Kleisli`, a singleton consisting of factory functions for that type alias, and an implicit syntax because the `Kleisli` isn't actually sufficient continues to look like a hack to me. 
If this is where the http4s server interface settles documentation is a must. Without it I speculate that few will understand it in its totality and thus most of the utility will be lost.
